gulpfile.js

```diff
@@ -191,3 +196,105 @@ gulp.task('docs', ['clean-docs'], function () {
     })
     .pipe(gulp.dest('docs'));
 });
+
+gulp.task('e2etest', function() {
+  var cmd = '--env default';
+  if(argv.browserstack) {
+    var browsers = require('./browsers.json');
+    var env = [];
+    var input = 'bs';
+    for(var key in browsers) {
+      if(key.substring(0, input.length) === input) {
+        env.push(key);
+      }
+    }
+
+    cmd = '--env default,' + env.join(',');
+  }
+
+  if(argv.browserstack) {
+    cmd = cmd + ' --config nightwatch.conf.js';
+  } else {
+    cmd = cmd + ' --config nightwatch.json';
+  }
+
+  if (argv.group) {
+    cmd = cmd + ' --group ' + argv.group;
+  }
+
+  cmd = cmd + ' --reporter ./test/spec/e2e/custom-reporter/pbjs-html-reporter.js';
+  return gulp.src('')
+    .pipe(shell('nightwatch ' + cmd));
+
+});
+
+gulp.task('e2etest-report', function(){
+  var browsers = require('./browsers.json');
+  var env = ['default'];
+  var input = 'bs';
+  for(var key in browsers) {
+    if(key.substring(0, input.length) === input) {
+      env.push(key);
+    }
+  }
+
+  //create new directory structure
+  var targetDestinationDir = './e2etest-report';
+  fs.rmrfSync(targetDestinationDir);
+  env.forEach(item => {
+    fs.mkdirpSync(targetDestinationDir + '/' + item);
+  });
+
+  //move xml files to newly created directory
+  var walker = fs.walk('./build/coverage/e2e/reports');
+  walker.on("file", function (root, stat, next) {
+    env.forEach(item => {
+      if(stat.name.search(item) !== -1) {
+        var src = root + '/' + stat.name;
+        var dest = targetDestinationDir + '/' + item + '/' + stat.name;
+        fs.copy(src, dest, {replace: true}, function(err) {
+          if(err) {
+            throw err;
+          }
+        });
+      }
+    });
+    next();
+  });
+
+  //run junit-viewer to read xml and create html
+  env.forEach(item => {
+    //junit-viewer --results="./custom-reports/chrome51" --save="./chrome.html"
+    var cmd = 'junit-viewer --results="' + targetDestinationDir + '/' + item + '" --save="' + targetDestinationDir + '/' + item +'.html"';
+    exec(cmd);
+  });
+
+  //create e2e-results.html
```

Comment:
Do we want to move some of the these utility function code into a separate file to keep gulpfile clear?